packages/minimongo/local_collection.js

```diff
@@ -443,43 +586,87 @@ export default class LocalCollection {
       }
     });
 
-    this._observeQueue.drain();
+    await this._observeQueue.drain();
 
     // If we are doing an upsert, and we didn't modify any documents yet, then
     // it's time to do an insert. Figure out what document we are inserting, and
     // generate an id for it.
     let insertedId;
     if (updateCount === 0 && options.upsert) {
       const doc = LocalCollection._createUpsertDocument(selector, mod);
-      if (! doc._id && options.insertedId) {
+      if (!doc._id && options.insertedId) {
         doc._id = options.insertedId;
       }
 
-      insertedId = this.insert(doc);
+      insertedId = await this.insertAsync(doc);
       updateCount = 1;
     }
 
-    // Return the number of affected documents, or in the upsert case, an object
-    // containing the number of affected docs and the id of the doc that was
-    // inserted, if any.
-    let result;
-    if (options._returnObject) {
-      result = {numberAffected: updateCount};
-
-      if (insertedId !== undefined) {
-        result.insertedId = insertedId;
-      }
-    } else {
-      result = updateCount;
+    return this.finishUpdate({
+      options,
+      insertedId,
+      updateCount,
+      callback,
+    });
+  }
+  // XXX atomicity: if multi is true, and one modification fails, do
+  // we rollback the whole operation, or what?
+  update(selector, mod, options, callback) {
+    if (! callback && options instanceof Function) {
+      callback = options;
+      options = null;
     }
 
-    if (callback) {
-      Meteor.defer(() => {
-        callback(null, result);
-      });
+    if (!options) {
+      options = {};
     }
 
-    return result;
+    const matcher = new Minimongo.Matcher(selector, true);
+
+    const qidToOriginalResults = this.prepareUpdate(selector);
+
+    let recomputeQids = {};
+
+    let updateCount = 0;
+
+    this._eachPossiblyMatchingDocSync(selector, (doc, id) => {
+      const queryResult = matcher.documentMatches(doc);
+
+      if (queryResult.result) {
+        // XXX Should we save the original even if mod ends up being a no-op?
+        this._saveOriginal(id, doc);
+        recomputeQids = this._modifyAndNotifySync(
+          doc,
+          mod,
+          queryResult.arrayIndices
+        );
+
+        ++updateCount;
+
+        if (!options.multi) {
+          return false; // break
+        }
+      }
+
+      return true;
+    });
+
+    Object.keys(recomputeQids).forEach(qid => {
+      const query = this.queries[qid];
+      if (query) {
+        this._recomputeResults(query, qidToOriginalResults[qid]);
+      }
+    });
+
+    this._observeQueue.drain();
+
```

Comment:
Aren't we forgetting to add the `upsert` scenario here? We are adding for the `updateAsync` but not for the "regular" one.